File: packages/nuget/src/Pattern.cs

namespace DesignPatterns.Catalog;

public class Pattern
{
    public Pattern(int id, string slug, string name, string category, string intent, string useWhen)
    {
        Id = id;
        Slug = slug;
        Name = name;
        Category = category;
        Intent = intent;
        UseWhen = useWhen;
    }

    public int Id { get; }
    public string Slug { get; }
    public string Name { get; }
    public string Category { get; }
    public string Intent { get; }
    public string UseWhen { get; }
}

File: web/src/types/pattern.ts
export type PatternCategory = 'Creational' | 'Structural' | 'Behavioral';

export interface Pattern {
  /** Position of the pattern in the classic Gang of Four catalog (1-23). */
  id: number;
  /** URL/DOM friendly identifier. */
  slug: string;
  name: string;
  category: PatternCategory;
  intent: string;
  useWhen: string;
  /** Ordered steps showing the pattern's typical collaboration flow. */
  flow: string[];
}

export const PATTERN_CATEGORIES: PatternCategory[] = [
  'Creational',
  'Structural',
  'Behavioral',
];
